modules/dnn/src/layers/reshape_layer.cpp

```diff
@@ -169,14 +169,13 @@ class ReshapeLayerImpl : public ReshapeLayer
     {
         for (size_t i = 0; i < inputs.size(); i++)
         {
-            Mat srcBlob = *inputs[i];
-            MatShape inputShape = shape(srcBlob), outShape = shape(outputs[i]);
+            Mat& srcBlob = *inputs[i];
+            Mat& dstBlob = outputs[i];
+            MatShape inputShape = shape(srcBlob);
 
             if (performReordering)
             {
-                Mat reordered_blob(inputShape, srcBlob.type());
-
-                float *dstData = reordered_blob.ptr<float>();
+                float *dstData = dstBlob.ptr<float>();
```

Comment:
In some cases this line will lead to dstData and srcData point on the same memory (because layer is in-place). In can cause wrong results because scrData will be erased by writing into dstData and then will be read again. There are few ways to correct this: 
1. Reorder weights in fully-connected layer in case when there is a Reshape->FC consequence in tensorflow model; Reordering was implemented to handle such cases (Reshape->FC)
2. Check if reordering is needed in getMemoryShapes() function and return shape of input as shape for internal necessary blob; also it's possible to always request internal blob